### ISSUE TITLE
[UX] Clean up account edit page with vertical tabs

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#3872]
**Description of the need**
The current account/user edit page feels rather cluttered.

![Screen Shot 2019-06-17 at 12 29 48 PM](https://user-images.githubusercontent.com/871421/59624205-c5b9f400-90fb-11e9-8834-73b0353f5f06.jpg)

**Proposed solution**
Would this be better with vertical tabs rather than a long list of fieldsets?

---

PR by @BWPanda: https://github.com/backdrop/backdrop/pull/2753

**Comment by docwilmot** [comment]
Yes!

**Comment by klonos** [comment]
Definitely yes. ...if decided otherwise, just noting that we do support fieldset summaries, so we can collapse those by default, and add those summaries.

**Comment by BWPanda** [comment]
Yes please! Consistency FTW :-)

**Comment by BWPanda** [comment]
I've made an initial [PR](https://github.com/backdrop/backdrop/pull/2753) that adds existing fieldsets to vertical tabs:

![user_form](https://user-images.githubusercontent.com/2385329/60203312-1399cf00-9855-11e9-9aab-65d251b6950c.jpg)

Let me know if I've missed any.

**Comment by laryn** [comment]
Great @BWPanda!

If you activate a language, the **Language Settings** could also be a tab:
![Screen Shot 2019-06-26 at 1 27 42 PM](https://user-images.githubusercontent.com/871421/60205219-440e4880-9816-11e9-9d3e-0b9fda4672da.jpg)

Personally I think it would be nice to have **Status** and **Roles** in the first/open vertical tab as well -- anyone else agree?

**Comment by BWPanda** [comment]
Aww, did you choose Estonian just for me? 😊

What should the first vertical tab be called?

**Comment by laryn** [comment]
Yep, specially for you. 😺 

Good question. To get a little bit radical here, what about putting everything else (username, password, email, status, roles, etc.) into a first/default tab called "Account settings"? Is that bad UI for some reason I'm not thinking of? (It seems like it would pull everything up nice and tight on the page).

**Comment by klonos** [comment]
I am making this a UX task, and proposing next bug release.

**Comment by klonos** [comment]
...while we're on it, can we please also make the blocked/active radio set more compact (so that there's even less scrolling on that page)? I was thinking either of the following:

- align the 2 options horizontally:

![Screen Shot 2019-06-27 at 11 03 51 am](https://user-images.githubusercontent.com/2423362/60225788-ac6b3180-98cb-11e9-85a8-09dddb53d70e.png)

- convert them to a single "Active" checkbox (experimented with also changing the label to "Account status" instead of just "Status")

![Screen Shot 2019-06-27 at 11 06 42 am](https://user-images.githubusercontent.com/2423362/60225790-b0974f00-98cb-11e9-9791-41c10d02e43f.png)

Also, merge the "Signature settings" and "Picture" tabs into a single "Personalization" tab (that's what we call these in `admin/config/people/settings`)

We can also merge "Locale settings" and "Language settings" into a "Language and regional settings" tab:

![Screen Shot 2019-06-27 at 11 00 30 am](https://user-images.githubusercontent.com/2423362/60225629-ebe54e00-98ca-11e9-9aee-dc3538c43655.png)

**Comment by BWPanda** [comment]
@klonos Your personalisation and language suggestions are good, but not sure I agree with the blocked/active ones...

If we add those options to a tab as @laryn suggested, that'll mean less vertical space anyway. And I think having a specified 'blocked' option means its obvious what it does, as opposed to simply un-checking 'active' which could be ambiguous...

As for @laryn's suggestion about putting all fields in vertical tabs, I think this has been suggested before/elsewhere (maybe even by me), but rejected as that's not how the UI's supposed to work. I'll see if I can find that issue.

**Comment by BWPanda** [comment]
Here you go. @klonos in #3171:

>

**Comment by BWPanda** [comment]
https://github.com/backdrop/backdrop-issues/issues/3894 will need to happen before we can add the `Status` and `Roles` settings to a fieldset/vertical tab.

**Comment by docwilmot** [comment]
> As for @laryn's suggestion about putting all fields in vertical tabs, I think this has been suggested before/elsewhere (maybe even by me), but rejected as that's not how the UI's supposed to work.
He then referenced https://www.drupal.org/docs/develop/user-interface-standards/vertical-tabs which says:


I think vertical tabs are perfect here for those same reasons in that article: everything there is often ignored and has defaults.


> #3894 will need to happen before we can add the Status and Roles settings to a fieldset/vertical tab.

Why group them, why not just give them their own tab each?

I like this issue because then the user edit page wont be that horrible mishmash of fields and pseudo fields. It'll be just custom fields and the vertical fieldset group.

**Comment by BWPanda** [comment]
@docwilmot I disagree about putting everything into vertical tabs. For sites without custom User fields, that'll mean there's nothing on the form _but_ vertical tabs, and my understanding of that Drupal docs page is that that's not the idea.

As for making `Status` and `Roles` separate tabs, I think that'll make too many tabs, especially with very few form elements in each.

**Comment by BWPanda** [comment]
I've pushed some changes, but not all. Still need to work on moving the languages and locale settings into the same tab, but as they come from different modules it's a bit more tricky... Making as `Needs work` for that reason.

**Comment by docwilmot** [comment]
> @docwilmot I disagree about putting everything into vertical tabs

Youre probably right about that. I'd like to see us stuffing as many fields as possible into verticals though.

But I think if https://github.com/backdrop/backdrop-issues/issues/3894 is blocking, we could baby step this and give the status and roles set their own vertical tabs. Its not too much; the content types have nine tabs.

**Comment by klonos** [comment]
#3894 will need to wait for 1.14, so perhaps get this in 1.13.3, and consider moving the status and roles into vtab part of #3894.

PS: I realised that it was me who se the milestone candidate 😅 ...so it'll have to be someone else to set the milestone if they agree.

**Comment by klonos** [comment]
> Still need to work on moving the languages and locale settings into the same tab...

You mean the same tab *group*. Right @BWPanda?

**Comment by BWPanda** [comment]
@klonos I meant a single tab, which is what I thought you mean by:

> We can also merge "Locale settings" and "Language settings" into a "Language and regional settings" tab

in https://github.com/backdrop/backdrop-issues/issues/3872#issuecomment-506098943

**Comment by klonos** [comment]
All good 👍 ...we're talking about the same thing 😄

**Comment by BWPanda** [comment]
[PR](https://github.com/backdrop/backdrop/pull/2753) updated to move regional and language settings into one tab. I think that's everything discussed above. Please test away!

**Comment by klonos** [comment]
Given that 1.14.0 is a round the corner, and the soft-blocking #3894 got merged for 1.14, I have switched the milestone suggestion to minor.

Testing things next...

**Comment by klonos** [comment]
Here's "some" feedback:

- We need status text for the vtabs, so that there's an overview w/o the user having to click through all of them to see the settings.
- Remove the "settings" bit, and have the vtabs labels be plain "Personalization" and "Regional and language"? I would expect such a trailing "settings" bit to be added to the fieldsets in `/admin/config/people/settings` instead (see my 1st point in the "brain dump" section that follows).
- "Signature settings" is shown in the "Personalization" vtab, even with the respective "Enable signatures" setting being disabled in `/admin/config/people/settings` 👎  
- If you disable both "Enable signatures" and "Enable user pictures" settings from  `/admin/config/people/settings`, the "Personalization" vtab is completely hidden 👍 
- If you disable "Enable user pictures" and enable "Enable signatures" in `/admin/config/people/settings`, the "Upload picture" widget is still shown in the "Personalization" vtab 👎 
- I don't like the "Signature settings" and "Picture" `h4`s within the vtab (especially because they are bolded, and draw too much attention); I believe that the field labels are enough. Same goes for the "Regional settings" and "Language settings" `h4`s.
- If you disable the "Users may set their own time zone" option in `/admin/config/regional/settings`, the "Regional and language settings" tab is still rendered (but empty) 👎 
- The help text for the timezone field reads "Select the desired local time and time zone ..."; should be only "time zone".
- I have enabled the Language module, and added a language in `/admin/config/regional/language`, but there was no language option shown in the "Regional and language settings" vtab. I then enabled the Locale module, which showed a radio list of languages, but the help text for it says "This account's default language for e-mails." ...is the language setting in the user profile all that we use it for? ...I also then enabled the Content Translation module, but that help text did not change. Aren't we using that setting to show content in the user's preferred language? As a novice site admin, I was confused; I can only clarify things either by experimentation, or by checking what the code does.
- I find the way that the "Current password" field is wedged between the username and email to be disruptive to the flow of things, but I cannot exactly put my finger on it at the moment; ...and that "Required if you want to change the E-mail address or Password below." help text makes me think that there is definitely somethings wrong if we need to be explaining the positions of fields within the form 🤔 ...perhaps we should have the order be username, then email, then new password, then current password(?).
- I was thinking that some sort of "configure this here" links in these vtabs would help. They would act as contextual links, and only be shown to site admins. The goal being to help them navigate to the respective pages elsewhere in the admin UI, where they can configure these settings.

## Brain dump (sorry 😅)

What follows below is the result of what ended up being a general UX/UI review of the specific page we are trying to overhaul here + the process and related pages that affect the elements being rendered in this form. Some of these may be pre-existing issues, and some may even be out of scope. These could be dealt with in follow-ups; just noting them here so that they don't get lost. So if unrelated, please ignore, and I will move them to separate issues (but perhaps we can sneak any trivial fixes in here 😅) ...here goes:

- Should we make the page "Account settings" in `/admin/config/people/settings` match the vtabs in the user account edit form? ...in other words, have separate vtabs for "Personalization", "Regional and language", and "URL alias pattern"
- Labels in some radio options and checkboxes in `/admin/config/people/settings` have trailing periods; I'm not sure if this is common practice, or specified in any coding or UI pattern standard, but I believe we don't do that.
- Help text in `/admin/config/people/settings` needs proper punctuation.
- Why don't we provide a proper image upload field for the default user picture in `/admin/config/people/settings`?
- Split the "Picture upload dimensions" to 2 separate "max height" / "max width" number fields.
- The text specified in the "Picture guidelines" text field in `/admin/config/people/settings` is being appended to the help text for the signature; I'd prefer it to be separate (bullet points perhaps?)
- I was thinking about the "When an email address is used as username, require a matching email address." setting in `/admin/config/people/settings`. If this is an option, then why not have it so that the email is always required, but the username is optional? If the user specifies a username, then that can be used; if not, then we internally store the email as the username.
- I was also thinking that perhaps the "Users may set their own time zone." setting + the dependent "Remind users at login if their time zone is not set." and "Time zone for new users" settings belong in `/admin/config/people/settings` instead of `/admin/config/regional/settings`.
- Replace the "Users with the _Select method for cancelling account_ or _Administer user accounts permissions_ can override this default method." with an actual permissions matrix, with only these 2 permissions.

I have more ideas/thoughts, but gonna stop now 😅

**Comment by klonos** [comment]
> - Why don't we provide a proper image upload field for the default user picture in `/admin/config/people/settings`?

I have created #4110 for this 😉

**Comment by BWPanda** [comment]
[PR](https://github.com/backdrop/backdrop/pull/2753) has been updated to address above comments/suggestions . Please let me know if I missed anything. Note that I didn't properly read through the 'brain dump', as it seems mostly out-of-scope for this issue.

**Comment by indigoxela** [comment]
This is what the PR provides:

![account-vertical-tabs](https://user-images.githubusercontent.com/842066/158325075-6b8fcb20-b426-4b34-ae6a-e38ee5307ee8.png)

**Comment by klonos** [comment]
> Note that I didn't properly read through the 'brain dump', as it seems mostly out-of-scope for this issue.

That's fine, I'll file follow-ups as/if needed 👍🏼 

Feedback after testing:
- the vtab label is `Region and language` instead of `Regional and language`, also the `and language` bit should only be added if the option for the language is available (if the Locale module is enabled)
- The help text says "Select the desired local time and time zone." although allowing only to select timezone - no option for time. We should fix that I think.

**Comment by BWPanda** [comment]
> the vtab label is `Region and language` instead of `Regional and language`

I did that because `Regional` sounded weird without the word `settings` after it...

> the `and language` bit should only be added if the option for the language is available

So the name of the tab will be 'Regional' if Locale isn't installed, 'Language' if it is but the setting for user timezones is disabled, and 'Regional and language' if both are available? Won't that be confusing? Imagine trying to write a how-to:

- Edit the user account
- Scroll down and click on the 'Regional' tab (or it might be called 'Language', or perhaps 'Regional and language'...)

> The help text says "Select the desired local time and time zone." although allowing only to select timezone - no option for time.

I looked at this, and I think the "local time" part refers to the time offset after the timezone:

![image](https://user-images.githubusercontent.com/2385329/158355771-e1e62649-a938-447f-869a-8a78076b9f75.png)

That was my interpretation anyway...

**Comment by olafgrabienski** [comment]
> So the name of the tab will be 'Regional' if Locale isn't installed, 'Language' if it is but the setting for user timezones is disabled, and 'Regional and language' if both are available? Won't that be confusing?

I agree. Also, the title of the page `admin/config/regional` is alway "Regional and language", even if *Locale* and *Language* are not enabled.

**Comment by BWPanda** [comment]
> Also, the title of the page `admin/config/regional` is alway "Regional and language"

Ah, didn't realise that. In that case I see the benefit in consistency and have updated the PR to change the tab name back to 'Regional and language'.

**Comment by indigoxela** [comment]
There's currently a conflict, that must be resolved. And is it just me, or does the Tugboat sandbox fail to start for others, too?

**Comment by BWPanda** [comment]
PR fixed and updated. Ready for reviewing again.

**Comment by stpaultim** [comment]
I like it. Much nicer than the current account edit page. Works for me, but I think this is important enough it would be nice to get another affirmation before changing the status.

**Comment by olafgrabienski** [comment]
Oh, too quick – after enabling Signatures and User pictures I noticed a strange behavior in the summary text of the "Personalization" tab. Here is what I did:

1. Edited the page of a user account
2. Wrote some text in the Signature field, tab summary: missing (should be `Signature`)
3. Uploaded a picture, summary: `Picture`, as expected for Picture, but `Signature`is still missing
4. Saved the account page
5. Edited the page again, summary: `Signature` (!)
  Should be `Signature` and `Picture`

**Comment by indigoxela** [comment]
I really don't want to be the party pooper here, but I do have one serious concern: the form structure of the form changes quite a bit, which may break contrib or custom hook_form_alter implementations.

Indicator for possible problems is that function locale_form_alter() and system_user_timezone(), which is used in system_form_user_profile_form_alter() and system_form_user_register_form_alter() needed an update to reflect the changes.
As these two would break, contrib and custom stuff might also break.

I'm not sure if this has already been considered here.

**Comment by jenlampton** [comment]
I love this improvement! 
I'm going to try the current PR on b.org where we've heavily altered the user edit form, to see what the potential impact might be.

**Comment by jenlampton** [comment]
Good news!  
We've *heavily* customized the edit form on backdropcms.org, but the things we changed are not these "default" options provided by core, so even though this PR alters those heavily, the changes we made to our edit form have not been disrupted: 

<img width="1205" alt="Screen Shot 2022-06-30 at 5 04 14 PM" src="https://user-images.githubusercontent.com/397895/176797741-437d9391-64db-4378-9558-628669a2d313.png">

We will need to add some CSS for vertical tabs (which we have not had on our front-end theme before) but other than that, things actually look pretty good!

One thought: We should probably move the contact form preferences above the regional settings, so that non-admins will see those options first, and be able to opt-out of getting email notifications without digging through the tabs.

**Comment by BWPanda** [comment]
PR updated with @jenlampton's suggestion to move Regional & Language tab below Contact Form tab.

**Comment by herbdool** [comment]
Code looks good. I believe too the pros outweigh the cons.

**Comment by herbdool** [comment]
There's a failing test but it doesn't look related: uninstall locale, and only on PHP 8.1

**Comment by herbdool** [comment]
In the meeting @quicksketch mentioned that many front end themes don't account for vertical tabs. What does this look like for users who see the front end theme when they visit their account page? I'm going to set this to needs testing until we have that.

**Comment by herbdool** [comment]
This is what it looks like in Basis. I used my phone with mobile mode and desktop mode in Firefox.

![Screenshot_20220827-081650](https://user-images.githubusercontent.com/1071483/187029948-b7347cb7-32a8-4544-99ba-c07ec86f48bd.png)
![Screenshot_20220827-081719](https://user-images.githubusercontent.com/1071483/187029950-63fb1174-01c0-425b-9e49-9a2c72fc4b03.png)

**Comment by herbdool** [comment]
Here's Bootstrap Lite. It's using the default layout which has a sidebar.

![Screenshot_20220827-084030](https://user-images.githubusercontent.com/1071483/187030676-14313324-c3d3-4d83-9484-26422210860c.png)

**Comment by bugfolder** [comment]
It's not clear to me that this is actually "cleaning up". Could this be made optional, via a config setting?

I tried this out on a local development version of a real site. For ordinary users, the only thing they can change that ended up in the vertical tabs is their time zone. Here's what it used to look like:

![before 2](https://user-images.githubusercontent.com/29214084/187035058-f663f516-b530-4c4b-9013-d0a1f25b0116.jpg)

and here's what it looks like with this change:

![after 2](https://user-images.githubusercontent.com/29214084/187035073-0377bde0-fb0e-4909-9fe7-cdb5c160da63.jpg)

It looks a bit cluttered now, and there's no real value in listing the current time zone setting immediately to the left of the dropdown.

For admin users, who can edit many more things, there are multiple vertical tabs, so potentially there's an organizational improvement. But with many possible roles, squeezing the role list now makes the long list even longer because the descriptions start to wrap. And also the Masquerade settings (which used to be at the bottom of the page) have now moved above the vertical tabs. Compare before (left) and after (right).
![both](https://user-images.githubusercontent.com/29214084/187035136-e0e370a9-9d6c-405a-8738-dd9615140052.jpg)

All things considered, I'd like at least the option to say "keep things as they currently are."

**Comment by herbdool** [comment]
A partial improvement would be out of scope of this issue but it would be to only convert fieldsets to vertical tabs when there are two or more. Could change https://github.com/backdrop/backdrop/blob/1.x/core/misc/vertical-tabs.js. I think that would be useful everywhere.

It's possible to make it an option I think. But I haven't looked too closely.

There will be some awkward exceptions for awhile such as with masquerade but eventually it would update too.

For the screenshots I posted it might also mean updating vertical tabs js file to improve the layout in narrow columns or default to mobile layout earlier. Might also be out of scope here.

**Comment by docwilmot** [comment]
I definitely think this cleans up the interface. The vertical fieldsets are meant to hide away fields that you dont need to really pay attention to, and there are usually a few of those for admins. I agree that screen width issues would be more a problem with vertical tabs generally rather than the fault of this PR. Looks good to me.

**Comment by docwilmot** [comment]
As mentioned by @indigoxela here https://github.com/backdrop/backdrop-issues/issues/3872#issuecomment-1169742408 though, we should be concerned about custom or contrib that is altering the user edit form. Have we discussed this in a Thursday meeting by any chance?

**Comment by herbdool** [comment]
I got a chance to look on a real desktop. Looks better. The breakpoints for vertical tabs can be a separate issue:

![2022-08-27 14 27 58 bd1 lndo site 1aedd92b0b35](https://user-images.githubusercontent.com/1071483/187043474-a068e6ac-e904-4f4f-87dd-f79fd023b39b.png)

**Comment by herbdool** [comment]
I created an issue and PR for only displaying vertical tabs if there is more than one: https://github.com/backdrop/backdrop-issues/issues/5747 so that might deal with one issue brought up above.

**Comment by klonos** [comment]
I also agree that this is a UX cleanup and improvement 👍🏼 ...excessive scrolling to get past infrequently-used settings that provide sensible defaults, in order to reach the submit buttons is bad UX and should be fixed where possible. Vertical tabs have been invented for that reason 😉 

I understand that things may not look ideal while contrib is trying to catch up with this change, but the UI in this form is not great as is either - especially when you have many contrib and core module add settings to it. I don't have any real-life site to demo how bad things can get, but here's what I have to live with each time I need to create user accounts in a D9 training site for GovCMS:

![image](https://user-images.githubusercontent.com/2423362/187047421-5f2dbc61-6d7c-44b2-9702-877eada5ee05.png)

...al I need is the first half of the page, and the rest are just "things that get in the way".

Thanks for opening the follow-up in $5747 @herbdool. I have also opened an issue in the queue of the contrib module Masquerade (of which you are a maintainer), to switch its settings to vtab one this lands in core: https://github.com/backdrop-contrib/masquerade/issues/6 😉

**Comment by stpaultim** [comment]
@herbdool Should this be RTBC?

**Comment by herbdool** [comment]
@stpaultim I think it could but waiting a little bit to get feedback.

1. Regarding an option to "turning this off" it might be good enough that someone could do it in a form alter to change the vertical tabs into fieldsets. Is that good enough or should it be easier? 
2. And regarding contrib and custom modules altering this form, those aren't blockers per se, since we don't make a guarantee, but would be good to notify maintainers after this is merged. (Such as https://github.com/backdrop-contrib/user_picture_field/blob/1.x-1.x/user_picture_field.module)

If people agree both are not blockers then we can rtbc this.

**Comment by bugfolder** [comment]
>  it might be good enough that someone could do it in a form alter to change the vertical tabs into fieldsets. Is that good enough or should it be easier?

Although i was the one who asked for the option to turn this off, I realize that adding such an option would not only require all the various core modules to check the option value when adding their fieldsets (to decide whether to set `#group` for the element), but also would require any contrib modules that add to the user edit page to check that option value when adding their element.

Or we could add a `hook_form_alter()` implementation that goes back and unsets any `#group` properties based on the option value, but we'd need to make sure our hook ran last; which might complicate matters with other contrib `hook_form_alter()`s that want to run last for their own unrelated reasons (e.g., rearranging order).

All seems a bit baroque to ask for, so I withdraw the request. Turning them back into fieldsets could be done as-needed, or by a small, specific contrib module.

An important part of the ovarall implementation would be https://github.com/backdrop/backdrop-issues/issues/5747, but I agree that it needn't be a blocker for this issue.

**Comment by stpaultim** [comment]
I think it's fine if those two concerns are not blockers. We can deal with them in follow-up issues.

**Comment by quicksketch** [comment]
I left a few comments in the PR: https://github.com/backdrop/backdrop/pull/2753

**Comment by quicksketch** [comment]
This is the last issue remaining in the 1.23.0 milestone. With the addition of https://github.com/backdrop/backdrop-issues/issues/5747, the situation where only a single fieldset is unnecessarily converted to a vertical tab is now avoided. We still have one more day (end of day, September 1); I think this might still be achievable.

**Comment by BWPanda** [comment]
I've committed @herbdool's suggestions (without really looking at them too much, so trusting others to review/approve).

**Comment by herbdool** [comment]
This looks good to me now. I tested again and still works the same. @quicksketch do you think it could squeak into this release still?

**Comment by quicksketch** [comment]
@klonos brought up that because this is a UX fix, it's potentially exempt from feature freeze entirely. Either way, I think this is also ready to go and I haven't made the preview release yet, so let's go ahead and pull it in. I merged https://github.com/backdrop/backdrop/pull/2753 into 1.x for 1.23.0! Yay!

https://github.com/backdrop/backdrop/commit/0d6dd2e6ce229d3523d54d8edf9fd357eb39566a by @BWPanda, @herbdool, @bugfolder, @laryn, @klonos, @docwilmot, @indigoxela, @olafgrabienski, @stpaultim, @jenlampton, and @quicksketch.

**Comment by jenlampton** [comment]
Just noting here that this is the first instance of adding vertical tabs to core that break our rule that "everything in a vertical tab can be safely ignored". By putting things in these tabs that shouldn't be ignored (like the user picture, in particular, but perhaps other things), we'll be breaking people's trust in our interfaces. 

I still think this is an improvement, but we're now on the slippery slope to module_filter madness! wheee.

**Comment by klonos** [comment]
@jenlampton I have always ignored user pictures in all Backdrop/Drupal installations - I don't think that it's a feature that shouldn't be ignored. Same for timezone and language. These things are either set-and-forget, or never-set-to-begin-with most of the time.

The only important things in the user profile form are the username/email/password - the rest can be safely ignored, and they are OK to be in vertical tabs I believe.

**Comment by docwilmot** [comment]
Hey, why dont we use Telemetry to see who's using user pictures?

**Comment by klonos** [comment]
> Hey, why dont we use Telemetry to see who's using user pictures?

😄 😛  ...we could. We can. Is this blocking any issues in our queue currently, where we are stuck on debates of what the current usage is and how a potentially BC-breaking change could affect existing sites?
